File: codes/ruby/chapter05/answer_A34.rb
# Input
n, x, y = gets.split.map(&:to_i)
a = gets.split.map(&:to_i)

grundy = Array.new(100001)
(0...100001).each do |i|
  transit = [false, false, false]
  if i >= x
    transit[grundy[i-x]] = true
  end
  if i >= y
    transit[grundy[i-y]] = true
  end

  if !transit[0]
    grundy[i] = 0
  elsif !transit[1]
    grundy[i] = 1
  else
    grundy[i] = 2
  end
end

# Output
xor_sum = 0
n.times do |i|
  xor_sum ^= grundy[a[i]]
end
if xor_sum >= 1
  puts "First"
else
  puts "Second"
end

File: codes/go/chapter05/answer_A34.go
package main

import (
	"fmt"
)

func main() {
	var N, X, Y int
	fmt.Scan(&N, &X, &Y)
	A := make([]int, N)
	for i := 0; i < N; i++ {
		fmt.Scan(&A[i])
	}

	grundy := make([]int, 100001)
	for i := 0; i < 100001; i++ {
		Transit := [3]bool{false, false, false}
		if i >= X {
			Transit[grundy[i-X]] = true
		}
		if i >= Y {
			Transit[grundy[i-Y]] = true
		}

		if Transit[0] == false {
			grundy[i] = 0
		} else if Transit[1] == false {
			grundy[i] = 1
		} else {
			grundy[i] = 2
		}
	}

	XOR_Sum := 0
	for i := 0; i < N; i++ {
		XOR_Sum ^= grundy[A[i]]
	}
	if XOR_Sum >= 1 {
		fmt.Println("First")
	} else {
		fmt.Println("Second")
	}
}
